doc/nrf/ug_bt_mesh_node_removal.rst

```diff
@@ -0,0 +1,198 @@
+===================
+
+Upload a new network key (or application key) to the nodes by sending the Config NetKey (AppKey) Update message to every node, including the node with the Configuration Client model.
+
+The snippet below demonstrates how to do this using the |NCS| API:
+
+.. code-block:: c
+
+    static uint8_t send_net_key_update(struct bt_mesh_cdb_node *node, void *user_data)
+    {
+        struct bt_mesh_cdb_subnet *subnet = user_data;
+        uint8_t status = 0;
+        int err;
+
+        err = bt_mesh_cfg_cli_net_key_update(node->net_idx, node->addr, subnet->net_idx,
+                                             subnet->keys[1].net_key, &status);
+        if (err || status) {
+            printk("Unable to update NetKey on %#x (err %d, status %u)\n", node->addr, err,
+                   status);
+        } else {
+            printk("Updated NetKey with index %u on %#x\n", subnet->net_idx, node->addr);
+        }
+
+        return BT_MESH_CDB_ITER_CONTINUE;
+    }
+
+    static int update_net_key(uint16_t net_idx, uint8_t *new_net_key)
+    {
+        struct bt_mesh_cdb_subnet *subnet;
+
+        subnet = bt_mesh_cdb_subnet_get(net_idx);
+        if (!subnet) {
+            printk("Unable to get subnet\n");
+            return -EINVAL;
+        }
+
+        /* Store the new network key in CDB. */
+        memcpy(subnet->keys[1].net_key, new_net_key, 16);
+        bt_mesh_cdb_subnet_store(subnet);
+
+        /* Send the new network key to each node. */
+        bt_mesh_cdb_node_foreach(send_net_key_update, subnet);
+        return 0;
+    }
+
+    ...
+
+    uint8_t new_net_key[16] = { ... };
+    update_net_key(BT_MESH_NET_PRIMARY, new_net_key);
+
+Switching to the new key
+========================
+
+After uploading the new key, configure the nodes to use it when sending messages.
+To do that, change the Key Refresh Phase to 2 by sending the Config Key Refresh Phase Set message with Transition field set to 0x2.
+
+Send the message to at least one node.
+The rest of the nodes will switch the Key Refresh Phase through Secure Network Beacon.
+But in this case, it may take a while for all nodes to eventually switch the phase since the time between 2 consecutive Secure Network Beacons is ~10 seconds.
+You can speed up this process by sending the Config Key Refresh Phase Set message to all nodes that need to be updated.
+
+After switching the Key Refresh Phase to 2, a node will decrypt messages using both old and new keys, but encrypt only using the new key.
+Therefore, when updating a network key, make sure to update the phase on the node running the Configuration Client model first, so it is able to decrypt the Config Key Refresh Phase Status message from other nodes when they switch to the new network key.
+
+The snippet below demonstrates how to set the Key Refresh Phase on all nodes using the |NCS| API:
+
+.. code-block:: c
+
+    static uint8_t send_key_refresh_phase_set(struct bt_mesh_cdb_node *node, void *user_data)
```

Comment:
It almost feels like we should push this into the shell app. Then there is no need of these snippets and the whole user guide can be explained using shell commands. If someone wants to do it using code, they can then refer to the shell app which will be more complete and functioning. This will also help in avoiding the maintenance of snippets here. That way guide will become more compact.

In the intro, you could say something like:
"This user guide shows how to execute the key refresh procedure using mesh shell commands. These commands utilize |NCS| APIs and can be referred if user wants to implement the procedure by themselves ..."

Reply:
Yes, I generally agree that it would be the correct to move this snippet to some sample that compiles and runs. However I'm not sure if moving it to `mesh_shell` or adding to `subsys/bluetooth/mesh/shell.c` commands would be correct. `subsys/bluetooth/mesh/shell.c` is basically an interface for mesh API. Moving this snippet there will make this code as de facto a standard and people will use without adjusting it to their needs. This code is still in a snippet state as error handling is not done correctly (`bt_mesh_cdb_node_foreach` doesn't allow to pass back error code, it has no error checking between steps). It could be part of `mesh_shell`, but then not clear where to add this command (it can't be added to `mesh` command, only `mesh models` allows to add commands dynamically). And I should also wait for the next upmerge before finalizing the PR. Not sure if all these hassle makes sense here...

Reply:
This will create two ways of doing things. One, potentially using shell, and one using this guide. If shell commands are not able to execute the Key Refresh procedure by invoking the individual commands, then something is missing in the config client APIs or CDB APIs, or mesh shell.

We could do this in shell afterward though and then replace (or remove) these parts from the user guide.

Reply:
Then I suggest to do this later.